controllers/login.js

```diff
@@ -81,6 +82,154 @@ router.post('/login/', (req, res, next) => {
 	}, req, res, errorSink);
 });
 
+router.post('/login/email', async (req, res) => {
+	const {
+		username,
+		password,
+		redirect,
+	} = req.body;
+
+	const payload = {
+		username,
+		password,
+	};
+
+	await authHelper.loginUser(req, res, 'local', payload, redirect, 'Email');
+});
+
+router.get('/login/email', (req, res) => {
+	const query = new URLSearchParams({
+		strategy: 'email',
+	});
+
+	if (req.query.redirect) {
+		query.append('redirect', redirectHelper.getValidRedirect(req.query.redirect));
+	}
+
+	res.redirect(redirectHelper.joinPathWithQuery('/login', query.toString()));
+});
+
+router.post('/login/ldap', async (req, res) => {
+	const {
+		username,
+		password,
+		schoolId,
+		system,
+		redirect,
+	} = req.body;
+
+	const systemId = system.split('//')[0];
+
+	const payload = {
+		username,
+		password,
+		systemId,
+		schoolId,
+	};
+
+	await authHelper.loginUser(req, res, 'ldap', payload, redirect, 'LDAP');
+});
+
+router.get('/login/ldap', (req, res) => {
+	const query = new URLSearchParams({
+		strategy: 'ldap',
+	});
+
+	if (req.query.redirect) {
+		query.append('redirect', redirectHelper.getValidRedirect(req.query.redirect));
+	}
+
+	res.redirect(redirectHelper.joinPathWithQuery('/login', query.toString()));
+});
+
+// eslint-disable-next-line consistent-return
+const redirectOAuth2Authentication = async (req, res, systemId, migration, redirect) => {
+	try {
+		const system = await api(req, { version: 'v3' }).get(`/systems/public/${systemId}`);
+
+		const { oauthConfig } = system;
+
+		if (!oauthConfig) {
+			return authHelper.handleLoginError(req, res, {
+				type: 'UNPROCESSABLE_ENTITY',
+				code: 422,
+			}, redirect);
+		}
+
+		const state = shortid.generate();
+
+		const authenticationUrl = authHelper.getAuthenticationUrl(oauthConfig, state);
+
+		req.session.oauth2State = {
+			state,
+			systemId,
+			systemName: system.displayName,
+			postLoginRedirect: redirect,
+			migration,
+		};
+
+		res.redirect(authenticationUrl.toString());
+	} catch (error) {
+		await authHelper.handleLoginError(req, res, error.error, redirect);
```

Comment:
what kind of error we try to catch? If we talk about server error, then the catch should be for api request, and not the whole logic